EIPS/eip-721.md

```diff
@@ -0,0 +1,312 @@
+    ///  queries about them do throw.
+    /// @return The non-zero address of the owner of deed `_deedId`, or `throw`
+    ///  if deed `_deedId` is not tracked by this contract
+    function ownerOf(uint256 _deedId) external view returns (address _owner);
+
+    /// @notice Count deeds tracked by this contract
+    /// @return A count of the deeds tracked by this contract, where each one of
+    ///  them has an assigned and queryable owner
+    function countOfDeeds() external view returns (uint256 _count);
+
+    /// @notice Count all deeds assigned to an owner
+    /// @dev Throws if `_owner` is the zero address, representing destroyed deeds.
+    /// @param _owner An address where we are interested in deeds owned by them
+    /// @return The number of deeds owned by `_owner`, possibly zero
+    function countOfDeedsByOwner(address _owner) external view returns (uint256 _count);
+
+    /// @notice Enumerate deeds assigned to an owner
+    /// @dev Throws if `_index` >= `countOfDeedsByOwner(_owner)` or if
+    ///  `_owner` is the zero address, representing destroyed deeds.
+    /// @param _owner An address where we are interested in deeds owned by them
+    /// @param _index A counter between zero and `countOfDeedsByOwner(_owner)`,
+    ///  inclusive
+    /// @return The identifier for the `_index`th deed assigned to `_owner`,
+    ///   (sort order not specified)
+    function deedOfOwnerByIndex(address _owner, uint256 _index) external view returns (uint256 _deedId);
+
+    // TRANSFER MECHANISM //////////////////////////////////////////////////////
+
+    /// @dev This event emits when ownership of any deed changes by any
+    ///  mechanism. This event emits when deeds are created (`from` == 0) and
+    ///  destroyed (`to` == 0). Exception: during contract creation, any
+    ///  transfers may occur without emitting `Transfer`.
+    event Transfer(address indexed from, address indexed to, uint256 indexed deedId);
+
+    /// @dev This event emits on any successful call to
+    ///  `approve(address _spender, uint256 _deedId)`. Exception: does not emit
+    ///  if an owner revokes approval (`_to` == 0x0) on a deed with no existing
+    ///  approval.
+    event Approval(address indexed owner, address indexed approved, uint256 indexed deedId);
+
+    /// @notice Approve a new owner to take your deed, or revoke approval by
+    ///  setting the zero address. You may `approve` any number of times while
+    ///  the deed is assigned to you, only the most recent approval matters.
+    /// @dev Throws if `msg.sender` does not own deed `_deedId` or if `_to` ==
+    ///  `msg.sender`.
+    /// @param _deedId The deed you are granting ownership of
+    function approve(address _to, uint256 _deedId) external payable;
+
+    /// @notice Become owner of a deed for which you are currently approved
+    /// @dev Throws if `msg.sender` is not approved to become the owner of
+    ///  `deedId` or if `msg.sender` currently owns `_deedId`.
```

Comment:
It can never happen that `msg.sender` is both approved to take ownership of `_deedId`, while also owning `_deedId`. The specification of `approve` prevents this. As such, it is not necessary to require a throw if `msg.sender` currently owns `_deedId`.

Reply:
After somebody performs `takeOwnership` then they are the owner. The standard does not require that the approval be reset at that time (such is an implementation detail). Therefore a compliant contract could implement this by making them the owner and keeping them as approved.

Reply:
Makes sense!

Reply:
Well, IMHO I think it would better to require approval reset. What is the purpose of leave approval set in this case?

Reply:
There is no purpose of leaving approval set. There is also no purpose of not leaving approval set. That's why it's an implementation detail.

Please take a look at f53cc2d, does this help?

Reply:
It helps, but I still think a strong requirement to clean the state would be better. Since approval/owns state track a deed state all states changes should be defined by the standard and not left to implementation avoiding nonsense states like approve a transfer  that would never happen is.

Reply:
Especially if the only standard flow is approve/takeownership. Even if implementation would choose to add different transfer functions, standard should require to always keep a correct clean state

Reply:
In terms of standard definition I would prefer to see the MUST word here to ensure state machine changes correctness preserved, but it's much much better than before